src/System.Reflection.Metadata/src/System/Reflection/Internal/MemoryBlocks/MemoryMappedFileBlock.cs

```diff
@@ -24,22 +25,19 @@ internal unsafe MemoryMappedFileBlock(IDisposable accessor, SafeBuffer safeBuffe
 
         ~MemoryMappedFileBlock()
         {
-            Dispose(false);
+            DisposeImpl();
         }
 
-        protected override void Dispose(bool disposing)
+        public override void Dispose()
         {
-            if (_safeBuffer != null)
-            {
-                _safeBuffer.ReleasePointer();
-                _safeBuffer = null;
-            }
-
-            if (_accessor != null)
-            {
-                _accessor.Dispose();
-                _accessor = null;
-            }
+            DisposeImpl();
+            GC.SuppressFinalize(this);
+        }
+
+        private void DisposeImpl()
+        {
+            Interlocked.Exchange(ref _safeBuffer, null)?.ReleasePointer();
```

Comment:
SafeHandle finalizer won't cleanup anything if there is pending reference. It assumes that there is other thread operating on the handle. The responsibility to cleanup is on whoever decrements the refcount to zero - the other thread in this case.